### PR TITLE
REPL: Wrap annotated expressions in a "resN" result val

### LINKED ISSUE #12292: Annotated expressions do not become a result in the REPL
## reproduction steps

using Scala 2.13.4,

```scala
scala> scala.#::.unapply(Stream(1))
       [warnings elided]
val res0: Option[(Int, Stream[Int])] = Some((1,Stream()))

scala> scala.#::.unapply(Stream(1)): @nowarn

scala> (scala.#::.unapply(Stream(1)): @nowarn)

scala> scala.#::.unapply(Stream(1)): @inline
       [warnings elided]

scala> (scala.#::.unapply(Stream(1)): @nowarn).isEmpty
val res1: Boolean = false
```

## problem

Annotated expressions are still expressions, and can still be used as expressions, so they should become a result (i.e. `resN`) in the REPL.

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes https://github.com/scala/bug/issues/12292

### CODE REVIEW

**Comment by SethTisue** [comment]
```
scala> scala.#::.unapply(Stream(1)): @nowarn
 val res1: Option[(Int, Stream[Int])] @scala.annotation.nowarn = Some((1,Stream()))
```

It seems wrong to me that `@nowarn` appears anywhere in the result.

`: @nowarn` isn't a type ascription, as per SLS 11.1; it shares a syntax with type ascriptions, but the meaning is to attach the annotation to the expression, it is not part of the type

and it doesn't make sense for the annotation to propagate from the expression to the definition (the `val`), which the transcript makes it appear to have done. the language doesn't work that way; when you write `val x = y: @foo`, the `@foo` annotation is only on the expression `y`, it isn't on the `val`. So it seems to me that the same ought to be true of `res1` here.

File: test/files/run/t12292.check [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+scala> (scala.#::.unapply(Stream(1)): @nowarn)
+val res2: Option[(Int, Stream[Int])] @scala.annotation.nowarn = Some((1,Stream()))
+
+scala> scala.#::.unapply(Stream(1)): @inline
+                 ^
+       warning: method unapply in object #:: is deprecated (since 2.13.0): Prefer LazyList instead
+                         ^
+       warning: value Stream in package scala is deprecated (since 2.13.0): Use LazyList instead of Stream
+                                      ^
+       warning: type Stream in package scala is deprecated (since 2.13.0): Use LazyList instead of Stream
```
Comment (NthPortal) [inline test/files/run/t12292.check:24]: why does this warning not occur for `res0`?

Comment (dwijnand) [reply test/files/run/t12292.check:24]: Looking at the arrow, it's pointing at `@inline`.  All that "it's not actually a type but using type ascription syntax" Seth was talking about... looks like it's causing this type deprecation to be emitted I think.

**Comment by lrytz** [comment]
> `: @nowarn` isn't a type ascription, as per SLS 11.1; it shares a syntax with type ascriptions, but the meaning is to attach the annotation to the expression

This is right. Just as a detail, in the implementation, annotated expressions (`Annotated` trees) are transformed by the type checker to `Typed(tree, tpe)` trees, which are type ascriptions. The annotation goes into the `tpe` as `AnnotatedType`.

> it doesn't make sense for the annotation to propagate from the expression to the definition

Yes, I agree. There is a variant of annotations that does that, but it's rather experimental. 

```
scala> class a extends annotation.TypeConstraint
class a

scala> object T { def f = 1: @a }
object T

scala> T.f
val res0: Int @a = 1
```

**Comment by lrytz** [comment]
> It seems wrong to me that `@nowarn` appears anywhere in the result.

Actually, this is also happening outside the REPL:

```
scala> class C { def f = 1: @inline }
class C

scala> typeOf[C].member(TermName("f")).info
val res0: $r.intp.global.Type = Int @inline
```

The method that widens the constant type `1` (computed by the type checker) to `Int` (assigned to the method) is here:
https://github.com/scala/scala/blob/f83a55b03b4017b3e9edb674e891b520026481a8/src/compiler/scala/tools/nsc/typechecker/Namers.scala#L1112

Maybe it should also drop annotations, I'm not sure what the impact would be.

In any case, references to `C.f` will have type `Int`, the annotation is dropped in `AsSeenFromMap`.

```
scala> (new C).f
[[syntax trees at end of                     typer]] // <console>
...
private[this] val res0: Int = new $line3.$read.INSTANCE.$iw.C{C}{(): C}(){C}.f{Int};
```

**Review by lrytz — APPROVED** [review]

**Pull request merged**